### DISCUSSION TITLE
`dagger develop` support for re-naming/re-sdk'ing existing module

### DISCUSSION THREAD

**Discussion by sipsma** [dagger/dagger#12437] [Ideas]
(Follow-up to https://github.com/dagger/dagger/pull/6533)

`dagger develop` currently only allows you to set the name/SDK of a module that does not already have one set.

It's possible to support changing the name/SDK too, but there's a few things to consider here:
* If you rename your module, you'd need to also update your code to use that name in the structs, but obviously much better would be if we did that automatically somehow
   * more thoughts from @helderco [here](https://github.com/dagger/dagger/pull/6533#issuecomment-1918962622)
* If you change your SDK, we can't translate your existing code into the new SDK lang, but it's also not clear if we should delete it (maybe fine, but it's scary to delete user files)

**Comment by rajatjindal** [comment]
Hi @sipsma 

I am starting to evaluate the requirements (and changes needed) for this.

**Comment by rajatjindal** [comment]
building on the suggestion from @helderco [here](https://github.com/dagger/dagger/pull/6533#issuecomment-1918962622), if we change code generator to generate the modules' `main.go` as follows:

Note: `Foo` is the name of the module and `InbuiltName` is a static name which will be constant across all modules.

 (`InbuiltName` is a random name I selected, we will change it to something more meaningful before merging.)

```
type Foo struct {
	InbuiltName
}

type InbuiltName struct{}

func (b *Foo) GetInbuiltName() *InbuiltName {
	return &b.InbuiltName
}

// Returns a container that echoes whatever string argument is provided
func (m *InbuiltName) ContainerEcho(stringArg string) *dagger.Container {
	return dag.Container().From("alpine:latest").WithExec([]string{"echo", stringArg})
}

// Returns lines that match a pattern in the files of the provided Directory
func (m *InbuiltName) GrepDir(ctx context.Context, directoryArg *dagger.Directory, pattern string) (string, error) {
	return dag.Container().
		From("alpine:latest").
		WithMountedDirectory("/mnt", directoryArg).
		WithWorkdir("/mnt").
		WithExec([]string{"grep", "-R", pattern, "."}).
		Stdout(ctx)
}
```

and then also update code generator to generation the `func invokeSrc` (in `dagger.gen.go` file) as follows:

```
...
return (*Foo).GetInbuiltName(&parent).ContainerEcho(stringArg), nil
...
```

instead of

```
return (*Foo).ContainerEcho(&parent, stringArg), nil
```

it seems to work as expected. and I am hoping will make renaming of the module easier. 

and when using this module as dependency, we are still able to do `dag.Foo()......` 🎉 

I am still working on trying out how other sdk will play out but I wanted to publish this update to receive any early feedback that folks might have. 

thanks
